File: retrocpu_test_framework/src/assemble_link.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { createRequire } from "node:module";
import { ASM_DIST, FRAMEWORK_BUILD } from "./repo.js";
import { expandIncludes, expandIncludesFromFile } from "./expand_includes.js";
import {
  checkpointId,
  isSyntheticCheckpointGlobal,
} from "./checkpoint.js";
import { defsToCdb, imageToIntelHex } from "./hex_cdb.js";
import { parseCdb } from "../../retrocpu_emu/src/code_test/cdb.js";
import type {
  AsmCpuType,
  AsmSource,
  AssembleLinkOptions,
  AssembleToFilesOptions,
  AssembledModule,
  LinkedCheckpoint,
  LinkedImage,
} from "./types.js";

const requireAsm = createRequire(import.meta.url);

type AssembleFn = (source: string, cpu: AsmCpuType) => unknown;

type WriteRelFn = (result: unknown, moduleName: string) => string;

type LinkRelsWithSdldFn = (
  relPaths: string[],
  options?: { wordAddrFixup?: boolean; workDir?: string; outName?: string },
) => {
  hexText: string;
  cdbText: string;
  image: Uint8Array;
  defs: Map<string, number>;
};

type AssembleResultLike = {
  symbols: Map<string, number>;
};

/**
 * retrocpu_asm の dist から関数を読む。未ビルドならエラー。
 * @returns assemble / writeRel / linkRelsWithSdld
 */
function loadAsmRuntime(): {
  assemble: AssembleFn;
  writeRel: WriteRelFn;
  linkRelsWithSdld: LinkRelsWithSdldFn;
} {
  const assemblerPath = path.join(ASM_DIST, "assembler.js");
  if (!fs.existsSync(assemblerPath)) {
    throw new Error(
      `retrocpu_asm が未ビルドです: ${assemblerPath}\n` +
        "`cd retrocpu_asm && npm run build` または `npm test`（test_framework が依存ビルドする）",
    );
  }
  const { assemble } = requireAsm(assemblerPath) as { assemble: AssembleFn };
  const { writeRel } = requireAsm(path.join(ASM_DIST, "relWriter.js")) as {
    writeRel: WriteRelFn;
  };
  const { linkRelsWithSdld } = requireAsm(
    path.join(ASM_DIST, "sdldLink.js"),
  ) as { linkRelsWithSdld: LinkRelsWithSdldFn };
  return { assemble, writeRel, linkRelsWithSdld };
}

/**
 * ソース指定からモジュール名を決める。
 * @param source 入力
 * @returns 大文字モジュール名
 */
function resolveModuleName(source: AsmSource): string {
  if ("text" in source) {
    return source.module.toUpperCase();
  }
  if (source.module) {
    return source.module.toUpperCase();
  }
  return path.basename(source.file, path.extname(source.file)).toUpperCase();
}

/**
 * ソース列に MAIN モジュールが含まれるか。
 * @param sources 入力
 * @returns MAIN があれば true
 */
export function sourcesHaveMain(sources: AsmSource[]): boolean {
  return sources.some((s) => resolveModuleName(s) === "MAIN");
}

/**
 * `_CODE` を指定ワードから始める MAIN スタブ。割り込み退避領域（0–7）と重ならないようにする。
 * @param orgWord `_CODE` 原点（ワード）
 * @param cpu 先頭 `.cpu` に書く CPU（既定 mn1613）
 * @returns スタブソース
 */
export function mn1613MainStub(
  orgWord: number,
  cpu: AsmCpuType = "mn1613",
): string {
  return [
    `\t.cpu\t${cpu}`,
    "\t.area\t_CODE\t\t(REL,CON)",
    `\t.org\t0x${orgWord.toString(16).toUpperCase()}`,
    "\t.global\t__TEST_FRAME_MAIN",
    "__TEST_FRAME_MAIN:",
    "\th",
    "",
  ].join("\n");
}

/**
 * `.asm` をアセンブルし sdld でリンクしてイメージ・HEX・CDB を返す。
 * MAIN が無い MN1613 では `_CODE` 原点スタブ（既定 0x0200）を入れる。
 * @param options ソースと CPU
 * @returns リンク結果
 */
export function assembleAndLink(options: AssembleLinkOptions): LinkedImage {
  const cpu: AsmCpuType = options.cpu ?? "mn1613";
  const hasMain = sourcesHaveMain(options.sources);
  const codeOrgWord =
    options.codeOrgWord !== undefined
      ? options.codeOrgWord
      : !hasMain && (cpu === "mn1613" || cpu === "mn1610")
        ? 0x0200
        : 0;
  const { assemble, writeRel, linkRelsWithSdld } = loadAsmRuntime();

  const workDir = fs.mkdtempSync(path.join(os.tmpdir(), "tf-sdld-"));
  const relPaths: string[] = [];
  const modules: AssembledModule[] = [];

  /**
   * 1 モジュールをアセンブルして .rel を書く。
   * @param sourceText ソース
   * @param moduleName モジュール名
   * @param sourcePath 表示用パス
   */
  const addModule = (
    sourceText: string,
    moduleName: string,
    sourcePath: string,
  ): void => {
    const result = assemble(sourceText, cpu) as AssembleResultLike;
    const relText = writeRel(result, moduleName);
    const relPath = path.join(workDir, `${moduleName}.rel`);
    fs.writeFileSync(relPath, relText, "utf8");
    relPaths.push(relPath);
    modules.push({
      module: moduleName,
      sourcePath,
      symbols: result.symbols,
    });
  };

  try {
    if (codeOrgWord > 0) {
      addModule(mn1613MainStub(codeOrgWord, cpu), "MAIN", "<test_frame_main>");
    }

    const sources = [...options.sources];
    const mainIdx = sources.findIndex((s) => resolveModuleName(s) === "MAIN");
    if (mainIdx > 0) {
      const [mainSrc] = sources.splice(mainIdx, 1);
      sources.unshift(mainSrc!);
    }

    for (const source of sources) {
      const moduleName = resolveModuleName(source);
      let sourceText: string;
      let sourcePath: string;
      if ("file" in source) {
        sourcePath = path.resolve(source.file);
        sourceText = expandIncludesFromFile(sourcePath);
      } else {
        sourcePath = `<inline:${moduleName}>`;
        sourceText = expandIncludes(
          source.text,
          source.fromDir ?? process.cwd(),
        );
      }
      addModule(sourceText, moduleName, sourcePath);
    }

    const linked = linkRelsWithSdld(relPaths, {
      workDir,
      wordAddrFixup: cpu !== "tms9995",
    });
    const globals = new Map<string, number>();
    const globalBytes = new Map<string, number>();
    for (const [name, byteAddr] of linked.defs) {
      if (isSyntheticCheckpointGlobal(name)) continue;
      if (/^__CP\$/i.test(name)) continue;
      const key = name.toUpperCase();
      globalBytes.set(key, byteAddr >>> 0);
      globals.set(key, byteAddr >>> 1);
    }

    const parsed = parseCdb(linked.cdbText);
    const checkpoints: LinkedCheckpoint[] = parsed.checkpoints.map((cp) => ({
      name: cp.name,
      serial: cp.serial,
      id: checkpointId(cp.name, cp.serial),
      byteAddr: cp.byteAddr >>> 0,
      wordAddr: cp.wordAddr >>> 0,
    }));

    return {
      image: linked.image,
      globals,
      globalBytes,
      hexText: linked.hexText || imageToIntelHex(linked.image),
      cdbText: linked.cdbText || defsToCdb(linked.defs),
      modules,
      checkpoints,
    };
  } finally {
    fs.rmSync(workDir, { recursive: true, force: true });
  }
}

/**
 * アセンブル／リンクし Intel HEX と CDB をファイルへ書く。
 * @param options ソースと出力パス
 * @returns リンク結果
 */
export function assembleToHexCdb(options: AssembleToFilesOptions): LinkedImage {
  const linked = assembleAndLink(options);
  const hexFile = path.resolve(options.hexFile);
  const cdbFile = path.resolve(options.cdbFile);
  fs.mkdirSync(path.dirname(hexFile), { recursive: true });
  fs.mkdirSync(path.dirname(cdbFile), { recursive: true });
  fs.writeFileSync(hexFile, linked.hexText, "utf8");
  fs.writeFileSync(cdbFile, linked.cdbText, "utf8");
  return linked;
}

/**
 * セッション用の既定 HEX / CDB パス。
 * @returns build/session.ihx と build/session.cdb
 */
export function defaultHexCdbPaths(): { hexFile: string; cdbFile: string } {
  return {
    hexFile: path.join(FRAMEWORK_BUILD, "session.ihx"),
    cdbFile: path.join(FRAMEWORK_BUILD, "session.cdb"),
  };
}

/**
 * リンク済みグローバルのワードアドレスを探す。
 * @param image リンク結果
 * @param name ラベル名（大文字小文字無視）
 * @returns ワードアドレス
 * @throws 見つからないとき
 */
export function lookupWordAddr(image: LinkedImage, name: string): number {
  const key = name.toUpperCase();
  const g = image.globals.get(key);
  if (g !== undefined) {
    return g;
  }
  throw new Error(`Global symbol not found: ${name}`);
}

File: retrocpu_test_framework_rs/tests/assemble_link.rs
//! assembleAndLink 統合テスト（fixture + sdld）。
//!
//! 根拠: `retrocpu_test_framework_ts/test/assemble_link.unit.ts`

use std::fs;
use std::path::PathBuf;

use retrocpu_test_framework_rs::{
    assemble_and_link, assemble_to_hex_cdb, find_sdld, lookup_byte_addr,
    parse_cdb, sources_have_main, AsmCpuType, AsmSource, AssembleLinkOptions,
    AssembleToFilesOptions, FrameworkError,
};

fn fixtures_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/fixtures")
}

fn fixture(name: &str) -> PathBuf {
    fixtures_dir().join(name)
}

fn sdld_available() -> bool {
    find_sdld().is_ok()
}

#[test]
fn main_stub_places_code_at_0x0200_without_main() -> Result<(), FrameworkError> {
    if !sdld_available() {
        eprintln!("skip: sdld not found");
        return Ok(());
    }
    let linked = assemble_and_link(AssembleLinkOptions {
        cpu: AsmCpuType::Mn1613,
        sources: vec![AsmSource::File {
            file: fixture("lib.asm"),
            module: None,
        }],
        code_org_word: None,
    })?;
    assert!(*linked.globals.get("GL_LIB_ENTRY").unwrap_or(&0) >= 0x0200);
    assert_eq!(linked.globals.get("__TEST_FRAME_MAIN"), Some(&0x0200));
    Ok(())
}

#[test]
fn main_with_lib_writes_hex_cdb() -> Result<(), FrameworkError> {
    if !sdld_available() {
        eprintln!("skip: sdld not found");
        return Ok(());
    }
    let out = tempfile::tempdir().expect("tempdir");
    let hex_file = out.path().join("assemble_link_fix.ihx");
    let cdb_file = out.path().join("assemble_link_fix.cdb");
    let linked = assemble_to_hex_cdb(AssembleToFilesOptions {
        link: AssembleLinkOptions {
            cpu: AsmCpuType::Mn1613,
            sources: vec![
                AsmSource::File {
                    file: fixture("main.asm"),
                    module: Some("MAIN".into()),
                },
                AsmSource::File {
                    file: fixture("lib.asm"),
                    module: None,
                },
            ],
            code_org_word: None,
        },
        hex_file: hex_file.clone(),
        cdb_file: cdb_file.clone(),
    })?;
    assert!(linked.globals.contains_key("GL_MAIN"));
    assert!(linked.globals.contains_key("GL_LIB_ENTRY"));
    assert!(hex_file.is_file());
    assert!(cdb_file.is_file());
    let cdb = parse_cdb(&fs::read_to_string(&cdb_file).expect("read cdb")).expect("parse cdb");
    let main = cdb.by_name.get("GL_MAIN").expect("GL_MAIN");
    assert_eq!(main.byte_addr % 2, 0);
    assert_eq!(
        main.word_addr,
        *linked.globals.get("GL_MAIN").expect("GL_MAIN word")
    );
    assert!(fs::read_to_string(&hex_file)
        .expect("read hex")
        .contains(":00000001FF"));
    Ok(())
}

#[test]
fn tms9995_no_main_stub() -> Result<(), FrameworkError> {
    if !sdld_available() {
        eprintln!("skip: sdld not found");
        return Ok(());
    }
    let linked = assemble_and_link(AssembleLinkOptions {
        cpu: AsmCpuType::Tms9995,
        sources: vec![AsmSource::Text {
            text: [
                "\t.cpu\ttms9995",
                "\t.area\t_CODE\t\t(REL,CON)",
                "\t.global\tGL_TMS_ENTRY",
                "GL_TMS_ENTRY:",
                "\tnop",
                "",
            ]
            .join("\n"),
            module: "LIBTMS".into(),
            from_dir: None,
        }],
        code_org_word: None,
    })?;
    assert_eq!(linked.cpu, AsmCpuType::Tms9995);
    assert!(!linked.globals.contains_key("__TEST_FRAME_MAIN"));
    assert_eq!(lookup_byte_addr(&linked, "GL_TMS_ENTRY").expect("byte"), 0);
    Ok(())
}

#[test]
fn sources_have_main_detects_main_module() {
    assert!(sources_have_main(&[AsmSource::File {
        file: fixture("main.asm"),
        module: Some("MAIN".into()),
    }]));
}
